File: packages/sdk/ruby/packages/admin/lib/edgebase_admin/functions_client.rb
# frozen_string_literal: true

module EdgebaseAdmin
  class FunctionsClient
    def initialize(http_client)
      @http = http_client
    end

    def call(path, method: "POST", body: nil, query: nil)
      normalized_path = "/functions/#{path.sub(%r{^/}, "")}"

      case method.to_s.upcase
      when "GET"
        @http.get(normalized_path, params: query)
      when "PUT"
        @http.put(normalized_path, body)
      when "PATCH"
        @http.patch(normalized_path, body)
      when "DELETE"
        @http.delete(normalized_path)
      else
        @http.post(normalized_path, body)
      end
    end

    def get(path, query: nil)
      call(path, method: "GET", query: query)
    end

    def post(path, body = nil)
      call(path, method: "POST", body: body)
    end

    def put(path, body = nil)
      call(path, method: "PUT", body: body)
    end

    def patch(path, body = nil)
      call(path, method: "PATCH", body: body)
    end

    def delete(path)
      call(path, method: "DELETE")
    end
  end
end

File: packages/sdk/kotlin/admin/src/main/kotlin/dev/edgebase/sdk/admin/FunctionsClient.kt
package dev.edgebase.sdk.admin

import dev.edgebase.sdk.core.HttpClient

data class FunctionCallOptions(
    val method: String = "POST",
    val body: Map<String, Any?>? = null,
    val query: Map<String, String>? = null
)

class FunctionsClient(
    private val httpClient: HttpClient
) {
    suspend fun call(path: String, options: FunctionCallOptions = FunctionCallOptions()): Any? {
        val normalizedPath = "/functions/$path"
        return when (options.method.uppercase()) {
            "GET" -> httpClient.get(normalizedPath, options.query)
            "PUT" -> httpClient.put(normalizedPath, options.body ?: emptyMap())
            "PATCH" -> httpClient.patch(normalizedPath, options.body ?: emptyMap())
            "DELETE" -> httpClient.delete(normalizedPath)
            else -> httpClient.post(normalizedPath, options.body ?: emptyMap())
        }
    }

    suspend fun get(path: String, query: Map<String, String>? = null): Any? =
        call(path, FunctionCallOptions(method = "GET", query = query))

    suspend fun post(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "POST", body = body))

    suspend fun put(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "PUT", body = body))

    suspend fun patch(path: String, body: Map<String, Any?> = emptyMap()): Any? =
        call(path, FunctionCallOptions(method = "PATCH", body = body))

    suspend fun delete(path: String): Any? =
        call(path, FunctionCallOptions(method = "DELETE"))
}
